lxd/cgroup/init.go

```diff
@@ -64,6 +64,204 @@ func (info *Info) Mode() string {
+	// MemorySwapUsage resource control
+	MemorySwapUsage
+	// MemorySwappiness resource control
+	MemorySwappiness
+	// NetPrio resource control
+	NetPrio
+	// Pids resource control
+	Pids
+)
+
+// SupportsVersion indicates whether or not a given cgroup control knob is
+// available and in which type of cgroup filesystem.
+func (info *Info) SupportsVersion(resource Resources) (Backend, bool) {
+	switch resource {
+	case Blkio:
+		val, ok := cgControllers["blkio"]
+		if ok && val == V1 {
+			return V1, ok
+		}
+		return Unavailable, false
+	case BlkioWeight:
+		val, ok := cgControllers["blkio.weight"]
+		if ok && val == V1 {
+			return V1, ok
+		}
+		return Unavailable, false
+	case CPU:
+		val, ok := cgControllers["cpu"]
+		if ok && val == V1 {
+			return V1, ok
+		}
+		return Unavailable, false
+	case CPUAcct:
+		val, ok := cgControllers["cpuacct"]
+		if ok && val == V1 {
+			return V1, ok
+		}
+		return Unavailable, false
+	case CPUSet:
+		val, ok := cgControllers["cpuset"]
+		if ok && val == V1 {
+			return V1, ok
+		}
+		return Unavailable, false
+	case Devices:
+		val, ok := cgControllers["devices"]
+		return val, ok
+	case Freezer:
+		val, ok := cgControllers["freezer"]
+		return val, ok
+	case Memory:
+		val, ok := cgControllers["memory"]
+		return val, ok
+	case MemoryMaxUsage:
+		val, ok := cgControllers["memory.max_usage_in_bytes"]
+		return val, ok
+	case MemorySwap:
+		val, ok := cgControllers["memory.memsw.limit_in_bytes"]
+		if ok {
+			return val, ok
+		}
+
+		val, ok = cgControllers["memory.swap.max"]
+		if ok {
+			return val, ok
+		}
+
+		return Unavailable, false
+	case MemorySwapMaxUsage:
+		val, ok := cgControllers["memory.memsw.max_usage_in_bytes"]
+		if ok && val == V1 {
+			return V1, ok
+		}
+		return Unavailable, false
+	case MemorySwapUsage:
+		val, ok := cgControllers["memory.memsw.usage_in_bytes"]
+		if ok {
+			return val, ok
+		}
+
+		val, ok = cgControllers["memory.swap.current"]
+		if ok {
+			return val, ok
+		}
+
+		return Unavailable, false
+	case MemorySwappiness:
+		val, ok := cgControllers["memory.swappiness"]
+		if ok && val == V1 {
+			return V1, ok
+		}
+		return Unavailable, false
+	case NetPrio:
+		val, ok := cgControllers["net_prio"]
+		if ok && val == V1 {
+			return V1, ok
+		}
+		return Unavailable, false
+	case Pids:
+		val, ok := cgControllers["pids"]
+		if ok && val == V1 {
+			return V1, ok
+		}
+		return Unavailable, false
+	}
+
+	return Unavailable, false
+}
+
+// Supports indicates whether or not a given cgroup control knob is available.
+// Note, we use "knob" instead of "controller" because this map holds
+// controllers as well as new features for a given controller, i.e. you can
+// have "blkio" which is a controller and "blkio.weight" which is a feature of
+// the blkio controller.
+func (info *Info) Supports(resource Resources, cgroup *CGroup) bool {
+	val, ok := info.SupportsVersion(resource)
+	if val == V2 && cgroup != nil && !cgroup.UnifiedCapable {
+		ok = false
+	}
+	return ok
+}
```

Comment:
Comment says "knob" but code says "resource" and the type of "resource" (singular) is "Resources" (plural).
This seems a bit confusing :)

Reply:
Types unless they're a slice, should always be singular.